packages/vhd-lib/src/createVhdDirectoryFromStream.js

```diff
@@ -0,0 +1,179 @@
+    offset: batOffset,
+    size: batSize,
+  })
+
+  // sometimes some parent locator are before the BAT
+  index.sort((a, b) => a.offset - b.offset)
+  while (index.length > 0) {
+    const item = index.shift()
+    const buffer = await read(item.offset, item.size)
+    if (item.type === 'bat') {
+      // found the BAT : read it and ad block to index
+      for (let blockCounter = 0; blockCounter < header.maxTableEntries; blockCounter++) {
+        const batEntrySector = buffer.readUInt32BE(blockCounter * 4)
+        // unallocated block, no need to export it
+        if (batEntrySector === BLOCK_UNUSED) {
+          continue
+        }
+        const batEntryBytes = batEntrySector * SECTOR_SIZE
+        // ensure the block is not before the bat
+        assert.ok(batEntryBytes >= batOffset + batSize)
+        index.push({
+          type: 'block',
+          id: blockCounter,
+          offset: batEntryBytes,
+          size: blockAndBitmapSize,
+        })
+      }
+      // sort again index to ensure block and parent locator are in the right order
+      index.sort((a, b) => a.offset - b.offset)
+    } else {
+      yield { ...item, buffer }
+    }
+  }
+
+  /**
+   * the second footer is at filesize - 512 , there can be empty spaces between last block
+   * and the start of the footer
+   *
+   * we read till the end of the stream, and use the last 512 bytes as the footer
+   */
+  const bufFooterEnd = await readLastSector(stream)
+  assert(bufFooter.equals(bufFooterEnd), 'footer1 !== footer2')
+}
+
+function readLastSector(stream) {
+  return new Promise((resolve, reject) => {
+    let bufFooterEnd = Buffer.alloc(0)
+    stream.on('data', chunk => {
+      if (chunk.length > 0) {
+        bufFooterEnd = cappedBufferConcat([bufFooterEnd, chunk], SECTOR_SIZE)
+      }
+    })
+
+    stream.on('end', chunk => {
+      if (chunk) {
+        bufFooterEnd = cappedBufferConcat([bufFooterEnd, chunk], SECTOR_SIZE)
+      }
+      resolve(bufFooterEnd)
+    })
+    stream.on('error', reject)
+  })
+}
+
+const buildVhd = Disposable.wrap(async function* (handler, path, inputStream, { concurrency }) {
+  const vhd = yield VhdDirectory.create(handler, path)
+  await asyncEach(
+    parse(inputStream),
+    async function (item) {
+      switch (item.type) {
+        case 'footer':
+          vhd.footer = item.footer
+          break
+        case 'header':
+          vhd.header = item.header
+          break
+        case 'parentLocator':
+          await vhd.writeParentLocator({ ...item, data: item.buffer })
+          break
+        case 'block':
+          await vhd.writeEntireBlock(item)
+          break
+        default:
+          throw new Error(`unhandled type of block generated by parser : ${item.type} while generating ${path}`)
+      }
+    },
+    {
+      concurrency,
+    }
+  )
+  await Promise.all([vhd.writeFooter(), vhd.writeHeader(), vhd.writeBlockAllocationTable()])
+})
+
+export const createVhdDirectoryFromStream = async function (
```

Comment:
For the future, when possible, use function declaration instead of function expression:
- it's more idiomatic
- ensure the `.name` property is correctly set (but it's often detected with modern JS engines)